WebDriverAgentLib/Utilities/FBClassChainQueryParser.m

```
@@ -0,0 +1,373 @@
+}
+
+@end
+
+
+@implementation FBStarToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"*"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[FBSplitterToken.class, FBOpeningBracketToken.class];
+}
+
++ (NSUInteger)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation FBSplitterToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"/"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[FBStarToken.class, FBClassNameToken.class];
+}
+
++ (NSUInteger)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation FBOpeningBracketToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"["];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[FBNumberToken.class];
+}
+
++ (NSUInteger)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation FBNumberToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  NSMutableCharacterSet *result = [NSMutableCharacterSet new];
+  [result formUnionWithCharacterSet:[NSCharacterSet decimalDigitCharacterSet]];
+  [result addCharactersInString:@"-"];
+  return result.copy;
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[FBClosingBracketToken.class];
+}
+
+@end
+
+
+@implementation FBClosingBracketToken
+
++ (NSCharacterSet *)allowedCharacters
+{
+  return [NSCharacterSet characterSetWithCharactersInString:@"]"];
+}
+
+- (NSArray<Class> *)followingTokens
+{
+  return @[FBSplitterToken.class];
+}
+
++ (NSUInteger)maxLength
+{
+  return 1;
+}
+
+@end
+
+
+@implementation FBClassChainElement
+
+- (id) initWithType:(XCUIElementType)type position:(NSInteger)position
+{
+  self = [super init];
+  if (self) {
+    _type = type;
+    _position = position;
+  }
+  return self;
+}
+
+@end
+
+
+@implementation FBClassChainQueryParser
+
++ (BOOL)tokenizationErrorWithIndex:(NSUInteger)index originalQuery:(NSString *)originalQuery error:(NSError **)error
```

Why not return an error here, if it is a error constructor?